### PR TITLE
Enable XMC1400 Kit for Arduino board into the Arduino IDE

### PR DESCRIPTION
**Pull request by boramonideep**
By creating this pull request you agree to the terms in CONTRIBUTING.md.
https://github.com/Infineon/.github/blob/master/CONTRIBUTING.md
--- DO NOT DELETE ANYTHING ABOVE THIS LINE ---

CONTRIBUTING.md also tells you what to expect in the PR process.

Description
This adds the XMC 1400 Kit for Arduino board into the XMC for Arduino BSP

Related Issue
#189 on github

Context
Please refer to internal ticket https://jirard.intra.infineon.com/browse/DESMAKERS-2490

### CODE REVIEW

**Comment by techpaul** [comment]
IN general good suggest a modification for simplicity now and in future maintainability.

CURRENTLY only the XMC1100 and XMC1400 Boot Kits have a GPIO pin, instead of continually adding TWO DEFINES to the pins_arduino.h for each board
~~~
#define HAS_GPIO_RESET
#define GPIO_RESET_PIN
~~~

Change Main.cpp and reset.c (and any other place to use 

~~~
#ifdef HAS_GPIO_RESET
~~~

Use GPIO_RESET_PIN to access which port and pin to use for any operations if needed now or in future..

This simplifies readablity and maintainability of code.

While at it remove from reset.c
~~~
#include <reset.h>
~~
add instead
~~~
#include <Arduino.h>
~~~

In reset.h  remove 
~~~
#include <Arduino.h>
~~~

This removes circular fetches and reset.h is mainly referenced in Arduino.h anyway.

Basically .cpp and .c files should include <Arduino.h> where as .h files should NOT reference it to avoid circular references and unnecessary fetches and disk activity. Majority of cores .h files are included by Arduino.h .

Anything that calls any cores .h file should be called from Arduino.h or from a .c/.cpp module that has ALREADY included Arduino.h FIRST or after gcc standard includes

**Comment by techpaul** [comment]
Has the information been put together for the addition to the Wiki as well ?

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+#define ERU0_0_IRQHandler IRQ3_Handler // RESET
+#define ERU0_0_IRQn IRQ3_IRQn
+
+#ifdef ARDUINO_MAIN //index is arduino pin count 
+// Mapping of digital pins and comments
+const XMC_PORT_PIN_t mapping_port_pin[] =
+    {
+    /* 0  */    {XMC_GPIO_PORT1, 2}, // RX                                        P1.2
+    /* 1  */    {XMC_GPIO_PORT1, 3}, // TX                                        P1.3
+    /* 2  */    {XMC_GPIO_PORT0 , 5}, // GPIO / LED2                              P0.5
```
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:130]: According to board manual AND standard UNO Shield R3 this should be int 0 P1.1 according to board and P1.4 according to manual

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+    /* 3  */    {XMC_GPIO_PORT1 , 1}, // External int 0 / PWM40-1 output          P1.1
```
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:131]: According to board manual AND standard UNO Shield R3 this should be int 1

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+#define SCU_1_IRQHandler IRQ1_Handler //RTC
+#define SCU_1_IRQn IRQ1_IRQn
+
+#define ERU0_0_IRQHandler IRQ3_Handler // RESET
+#define ERU0_0_IRQn IRQ3_IRQn
+
+#ifdef ARDUINO_MAIN //index is arduino pin count 
+// Mapping of digital pins and comments
+const XMC_PORT_PIN_t mapping_port_pin[] =
```
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:126]: Many pins do not match Board legend and Board manual and standard Arduino Uno Shield R3

Example Arduino pin 2 is P0.5 here but in manual and board it is P1.4

Many other mismatches

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    /* 17  */   {XMC_GPIO_PORT2 , 9}, // A2 / ADC Input                           P2.9 (INPUT ONLY)
+    /* 18  */   {XMC_GPIO_PORT2 , 10}, // A3 / ADC Input                          P2.10
+    /* 19  */   {XMC_GPIO_PORT2 , 1}, // A4  / I2C Data / Address SDA             P2.1
+    /* 20  */   {XMC_GPIO_PORT2 , 0}, // A5  / I2C Clock SCL                      P2.0
+    /* 21  */   {XMC_GPIO_PORT2 , 4}, // RESET input ( DO NOT USE as GPIO )       P2.4 (INPUT ONLY)
+    /* 22 */    {XMC_GPIO_PORT2 , 11}, // GPIO / AD_AUX Additional Pin            P2.11
+    /* 23 */    {XMC_GPIO_PORT2 , 7}, // GPIO / AD_AUX Additional Pin             P2.7
+    /* 24 */    {XMC_GPIO_PORT2 , 5}, // GPIO / AD_AUX Additional Pin             P2.5
+    /* 25 */    {XMC_GPIO_PORT2 , 2}, // GPIO / AD_AUX Additional Pin             P2.2
+    /* 26 */    {XMC_GPIO_PORT1 , 4}, // External int 1                           P1.4
```
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:154]: Duplicate of Pin 2

**Review by techpaul — CHANGES_REQUESTED** [review]
See comments in discussion and on files

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+#define LED1        13
+#define LED2        26
+#define LED3        27
+#define LED_BUILTIN LED1
+
+#define EXT_INTR_0  3
+#define EXT_INTR_1  26
+
+#define digitalPinToInterrupt(p)    ((p) == 3 ? 0 : ((p) == 26 ? 1 : NOT_AN_INTERRUPT))
```
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:98]: Int 0 and 1 should be on pins 2 and 3

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    /* 18  */   {XMC_GPIO_PORT2 , 10}, // A3 / ADC Input                          P2.10
+    /* 19  */   {XMC_GPIO_PORT2 , 1}, // A4  / I2C Data / Address SDA             P2.1
+    /* 20  */   {XMC_GPIO_PORT2 , 0}, // A5  / I2C Clock SCL                      P2.0
+    /* 21  */   {XMC_GPIO_PORT2 , 4}, // RESET input ( DO NOT USE as GPIO )       P2.4 (INPUT ONLY)
+    /* 22 */    {XMC_GPIO_PORT2 , 11}, // GPIO / AD_AUX Additional Pin            P2.11
+    /* 23 */    {XMC_GPIO_PORT2 , 7}, // GPIO / AD_AUX Additional Pin             P2.7
+    /* 24 */    {XMC_GPIO_PORT2 , 5}, // GPIO / AD_AUX Additional Pin             P2.5
+    /* 25 */    {XMC_GPIO_PORT2 , 2}, // GPIO / AD_AUX Additional Pin             P2.2
+    /* 26 */    {XMC_GPIO_PORT1 , 4}, // External int 1                           P1.4
+    /* 27 */    {XMC_GPIO_PORT1 , 5} // LED3                                      P1.5
```
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:155]: Add comment ONLY drives LED

**Review by techpaul — CHANGES_REQUESTED** [review]
Further issues noted see my wiki on this and read it carefully including suggested pin number
https://github.com/techpaul/XMC-for-Arduino/wiki/XMC1400-Arduino-Kit-Pinout-and-Issues

File: libraries/SPI/src/utility/xmc_spi_conf.h [hunk 7]
```diff
@@ -60,6 +60,10 @@ extern XMC_SPI_t XMC_SPI_0;
 #define NUM_SPI  			1
 extern XMC_SPI_t XMC_SPI_0;
 
+#elif defined(XMC1400_Arduino_Kit)
+#define NUM_SPI  			1
+extern XMC_SPI_t XMC_SPI_0;
+
 #elif defined(XMC4400_Platform2GO)
```
Comment (jaenrig-ifx) [inline libraries/SPI/src/utility/xmc_spi_conf.h:67]: It feels all these sections of if else are including the same number of SPIs and same instance. Maybe they can be in one if block?

**Review by jaenrig-ifx — APPROVED** [review]
I think all these pins potential issues are due to the typos in documentation. 
All rest looks fine to me prior on target testing 👍

**Comment by techpaul** [comment]
@jaenrig-ifx @boramonideep

I have a physical board here and I would not even attempt at testing here.

Side note please do not use INTERNAL resources to Infineon on Github as per initial pull request.

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    /* 13  */   {XMC_GPIO_PORT0 , 3}, // SPI-SCK                                  P0.3
+    /* 14  */   {XMC_GPIO_PORT2 , 3}, // AREF                                     P2.3 (INPUT ONLY)
+    /* 15  */   {XMC_GPIO_PORT2 , 6}, // A0 / ADC Input                           P2.6 (INPUT ONLY)
+    /* 16  */   {XMC_GPIO_PORT2 , 8}, // A1 / ADC Input                           P2.8 (INPUT ONLY)
+    /* 17  */   {XMC_GPIO_PORT2 , 9}, // A2 / ADC Input                           P2.9 (INPUT ONLY)
+    /* 18  */   {XMC_GPIO_PORT2 , 10}, // A3 / ADC Input                          P2.10
+    /* 19  */   {XMC_GPIO_PORT2 , 1}, // A4  / I2C Data / Address SDA             P2.1
+    /* 20  */   {XMC_GPIO_PORT2 , 0}, // A5  / I2C Clock SCL                      P2.0
+    /* 21  */   {XMC_GPIO_PORT2 , 4}, // RESET input ( DO NOT USE as GPIO )       P2.4 (INPUT ONLY)
+    /* 22 */    {XMC_GPIO_PORT2 , 11}, // GPIO / AD_AUX Additional Pin            P2.11
```
Comment (djaumann) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:150]: I would number the AD_AUX here also in the comment, so AD_AUX_1 for this one here :)

**Review by djaumann — APPROVED** [review]
I checked the hardware as well and did not have any issues, only the mismatching of the silkscreen, which you've addressed already and the comment I left you regarding the AD_AUX or AUX_AD

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 2 above)
Comment (boramonideep) [reply variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:131]: An inconsistency was discovered between the Board Legend(silkscreen) and the wiring. The pins P0.5 and P1.4 have been swapped. Pin 2 on the board is actually P0.5, while P1.4 is 26.  

Hence P1.1 (pin 3) had to be named as INT0, while P1.4 (pin 26) as INT1.

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 3 above)
Comment (boramonideep) [reply variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:126]: As mentioned above, the interrupts had to be named differently. Please let us know of any other mismatches.

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 4 above)
Comment (boramonideep) [reply variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:154]: Already mentioned above

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 5 above)
Comment (boramonideep) [reply variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:98]: The interrupts had to named differently as mentioned above.

File: cores/WInterrupts.c [hunk 9]
```diff
@@ -145,6 +145,12 @@ void attachInterrupt(uint32_t interrupt_num, interrupt_cb_t callback, uint32_t m
             XMC_USIC_CH_Enable(XMC_USIC0_CH0);
             XMC_USIC_CH_SetInputSource(XMC_USIC0_CH0, XMC_USIC_CH_INPUT_DX5, USIC0_C0_DX5_P1_4);
             XMC_USIC_CH_SetInputSource(XMC_USIC0_CH0, XMC_USIC_CH_INPUT_DX2, USIC0_C0_DX2_DX5INS);
+#endif
```
[the remark this answers is not in the document]
Comment (boramonideep) [reply cores/WInterrupts.c:148]: Yes, edited in code to use the same block for both boards.

**Comment by techpaul** [comment]
@boramonideep  @jaenrig-ifx 
All your fixes are BACK TO FRONT

To follow the Hardware and Software Arduino pin referenceing WHATEVER the ID says Pin 2 and 3 MUST be interrupts NOT arduino pin 26.

This means om **ANY** board with the same connectors especially the XMC boards the same shield and SOFTWARE will work the same **WITHOUT MODIFICATION** !!

By having interrupt on Arduino pin 26 you have BROKEN the Arduino Uno Shieled R3 compatability. The idents are the LEAST of your worries. The pin order in software to match order on connectors matter regardless of ident.

You change the definition of pin 2 and 3 to match what SHOULD happen NOT the ident or amy other weird method.

In Arduino **NOBODY** use P1.4 referencing, they use Arduino pin wiring definitions

So the ident is wrong, you obviously understand little about the arduino structure

I also agree with @djaumann ) about putting comments about AD_AUX1 to AD_AUX_n

Also as commented above you need to instigate a board agnostic method of reset pin handling as in

#define HAS_GPIO_RESET
#define GPIO_RESET_PIN

This is a minimalistic approach to adding support, as in it works if you stand on one leg with your arm behind your back singing "Yankee doodle dandy"

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+extern const uint8_t NUM_INTERRUPT;
+extern const uint8_t NUM_ANALOG_INPUTS;
+#define NUM_LEDS            3
+#define NUM_SERIAL          1
+#define NUM_TONE_PINS       4
+#define NUM_TASKS_VARIANT   8
+
+// Indicate unit has RTC/Alarm
+#define HAS_RTC             1
+
```
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:56]: Need to change to board agnostic method of gpio reset pin
#define HAS_GPIO_RESET
#define GPIO_RESET_PIN

See conversation on pull request

File: cores/Main.cpp [hunk 11]
```diff
@@ -41,8 +41,9 @@ int main(void)
  */
 wiring_time_init();
 wiring_analog_init();
-// Initialize the reset pin for the XMC1100 Boot Kit series
-#ifdef XMC1100_Boot_Kit
+// Initialize the reset pin for the XMC1100 Boot Kit series and XMC1400 Kit for Arduino as they are based on Arduino form-factor
+// Hence, a dedicated reset pin is required. 
+#if defined(XMC1100_Boot_Kit) || defined(XMC1400_Arduino_Kit)
```
Comment (techpaul) [inline cores/Main.cpp:46]: Change to board agnostic method here using define in pins_arduino.h (also on XMC1100 Boot Kit)

#define HAS_GPIO_RESET

so 
#ifdef HAS_GPIO_RESET

File: cores/reset.c [hunk 12]
```diff
@@ -26,7 +26,7 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.   
  */
 #include <reset.h>
-#ifdef XMC1100_Boot_Kit
+#if defined(XMC1100_Boot_Kit) || defined(XMC1400_Arduino_Kit)
```
Comment (techpaul) [inline cores/reset.c:29]: Change to board agnostic method here using define in pins_arduino.h (also on XMC1100 Boot Kit)

#define HAS_GPIO_RESET

so
#ifdef HAS_GPIO_RESET

File: libraries/SPI/src/utility/xmc_spi_conf.h (hunk 7 above)
Comment (techpaul) [reply libraries/SPI/src/utility/xmc_spi_conf.h:67]: Or use NUM_SPI defined in pins_arduino.h to reduce core modifications

File: libraries/Wire/src/utility/xmc_i2c_conf.h [hunk 13]
```diff
@@ -64,6 +64,10 @@ extern XMC_I2C_t XMC_I2C_1;
 #define NUM_I2C  1
 extern XMC_I2C_t XMC_I2C_0;
 
+#elif defined(XMC1400_Arduino_Kit)
```
[the remark this answers is not in the document]
Comment (techpaul) [reply libraries/Wire/src/utility/xmc_i2c_conf.h:67]: My comment of NUM_I2C in pins_arduino.h as per NUM_SPI

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 8 above)
Comment (techpaul) [reply variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:150]: Agreed preferably with also an A6 to A9 definition AS WELL

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 4 above)
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:154]: NOT RESOLVED back to front modification that **BREAKS** Arduino comapatibility even within Infineon board set

Board ID is the LOWEST level of definition, not the master

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 5 above)
Comment (techpaul) [inline variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:98]: See Pull request conversation NOT FIXED BREAKS Arduino compatability.

**Review by techpaul — CHANGES_REQUESTED** [review]
Still things BROKEN and still things that need better implementation

File: libraries/SPI/src/utility/xmc_spi_conf.h (hunk 7 above)
Comment (boramonideep) [reply libraries/SPI/src/utility/xmc_spi_conf.h:67]: The implementation (C file) has a dedicated section for the XMC 1400 Kit for Arduino since most of the settings are unique to the board. Hence, a dedicated block in the H file to maintain similarity to the C file.

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 10 above)
Comment (boramonideep) [reply variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:56]: Currently, only two boards XMC 1100 Boot Kit and XMC 1400 Kit for Arduino have dedicated reset pins. Such a boad agnostic approach may be added when more boards with dedicated resets are added to the core.

File: cores/Main.cpp (hunk 11 above)
Comment (boramonideep) [reply cores/Main.cpp:46]: Currently, only two boards XMC 1100 Boot Kit and XMC 1400 Kit for Arduino have dedicated reset pins. Such a boad agnostic approach may be added when more boards with dedicated resets are added to the core.

Comment (techpaul) [reply cores/Main.cpp:46]: NO **NOW** is the **TIME** to do it BEFORE another board comes along

Obviously you think of this as a student exercise and everybody else is going to have to TIDY UP your **mess** later

Not resolved someone trying to do bare minimum to pass an exam

File: cores/reset.c (hunk 12 above)
Comment (techpaul) [reply cores/reset.c:29]: Same as before when you change from ONE to more you put in a better solution AT THAT point

File: libraries/SPI/src/utility/xmc_spi_conf.h (hunk 7 above)
Comment (techpaul) [reply libraries/SPI/src/utility/xmc_spi_conf.h:67]: Obviously you have a MISUNDERSTANDING about maintainability and readability.

The point about pins_arduino.h, is to SPECIFY the hardware specific definitions and flags THERE not in several other files.

**Comment by boramonideep** [comment]
> @boramonideep @jaenrig-ifx All your fixes are BACK TO FRONT
> 
> To follow the Hardware and Software Arduino pin referenceing WHATEVER the ID says Pin 2 and 3 MUST be interrupts NOT arduino pin 26.
> 
> This means om **ANY** board with the same connectors especially the XMC boards the same shield and SOFTWARE will work the same **WITHOUT MODIFICATION** !!
> 
> By having interrupt on Arduino pin 26 you have BROKEN the Arduino Uno Shieled R3 compatability. The idents are the LEAST of your worries. The pin order in software to match order on connectors matter regardless of ident.
> 
> You change the definition of pin 2 and 3 to match what SHOULD happen NOT the ident or amy other weird method.
> 
> In Arduino **NOBODY** use P1.4 referencing, they use Arduino pin wiring definitions
> 
> So the ident is wrong, you obviously understand little about the arduino structure
> 
> I also agree with @djaumann ) about putting comments about AD_AUX1 to AD_AUX_n
> 
> Also as commented above you need to instigate a board agnostic method of reset pin handling as in
> 
> #define HAS_GPIO_RESET #define GPIO_RESET_PIN
> 
> This is a minimalistic approach to adding support, as in it works if you stand on one leg with your arm behind your back singing "Yankee doodle dandy"

Thanks for the comment. Since now P0.5 appears at the Pin 2 location, it was tried to route INT0 through it but in vain. Hence, pin 3 (P1.1) had to named INT0 and pin 26 (P1.4) as INT1. We understand that this is not compatible with the Arduino Uno Rev3 Form Factor, but this was the only way forward.

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 4 above)
Comment (boramonideep) [reply variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:154]: Replied below

File: variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h (hunk 5 above)
Comment (boramonideep) [reply variants/XMC1400/config/XMC1400_Arduino_Kit/pins_arduino.h:98]: Replied below

**Comment by techpaul** [comment]
@boramonideep  @jaenrig-ifx 

I see you have merged this into develop I do hope someone gets time to tidy up this mess and laziness. I was thing of using the board on a project as obviously the hardware and documentation were released before proper checks and this has now been rushed through I will have to use different micros/board on a paying project.

I do not have time at present to fix your mess
